src/v/cloud_storage/remote_partition.cc

```diff
@@ -0,0 +1,641 @@
+                    if (
+                      !result
+                      && result.error()
+                           == storage::parser_errc::end_of_stream) {
+                        vlog(
+                          _ctxlog.debug, "EOF error while reading from stream");
+                        _reader->config().next_offset_redpanda
+                          = _reader->max_rp_offset() + model::offset(1);
+                        // Next iteration will trigger transition in
+                        // 'maybe_reset_reader'
+                        continue;
+                    } else if (!result) {
+                        vlog(_ctxlog.debug, "Unexpected error");
+                        throw std::system_error(result.error());
+                    }
+                    // empty result will also be propagated here
+                    data_t d = std::move(result.value());
+                    co_return storage_t{std::move(d)};
+                }
+            }
+        } catch (const ss::gate_closed_exception&) {
+            vlog(
+              _ctxlog.debug,
+              "gate_closed_exception while reading from remote_partition");
+            _it = _partition->_segments.end();
+            _reader = {};
+        }
+        vlog(
+          _ctxlog.debug,
+          "EOS reached {} {}",
+          static_cast<bool>(_reader),
+          is_end_of_stream());
+        co_return storage_t{};
+    }
+
+    void print(std::ostream& o) override {
+        o << "cloud_storage_record_batch_reader";
+    }
+
+private:
+    // Initialize object using remote_partition as a source
+    void initialize_reader_state(const log_reader_config& config) {
+        vlog(_ctxlog.debug, "record_batch_reader_impl initialize reader state");
+        auto lookup_result = find_cached_reader(config);
+        if (lookup_result) {
+            auto&& [reader, it] = lookup_result.value();
+            _reader = std::move(reader);
+            _it = it;
+            return;
+        }
+        vlog(
+          _ctxlog.debug,
+          "record_batch_reader_impl initialize reader state - segment not "
+          "found");
+        _it = _partition->_segments.end();
+        _reader = {};
+    }
+
+    struct cache_reader_lookup_result {
+        std::unique_ptr<remote_segment_batch_reader> reader;
+        remote_partition::segment_map_t::iterator iter;
+    };
+
+    std::optional<cache_reader_lookup_result>
+    find_cached_reader(const log_reader_config& config) {
+        if (!_partition || _partition->_segments.empty()) {
+            return std::nullopt;
+        }
+        auto it = _partition->_segments.lower_bound(config.start_offset);
+        if (it == _partition->_segments.end()) {
+            it = std::prev(it);
+        }
+        while (it->first > config.start_offset
+               && it != _partition->_segments.begin()) {
+            // scan back until the matching segment is found
+            it = std::prev(it);
+        }
```

Comment:
it looks like this entire block is equivalent to the following using upper_bound without the loop. but i haven't worked through it precisely

```
auto it = _partition->_segments.upper_bound(config.start_offset);
if (it != _partition->_segments.begin()) {
   it = std::prev(it);
}
```